django/utils/deprecation.py

```diff
@@ -83,6 +85,187 @@ def __new__(cls, name, bases, attrs):
+
+    The @deprecate_posargs decorator adds a small amount of overhead. In most cases
+    it won't be significant, but use with care in performance-critical code paths.
+    """
+
+    message_template_singular = (
+        "Passing positional argument {remapped_names} to {func_name}()"
+        " is deprecated. Change it to a keyword arg."
+    )
+    message_template_plural = (
+        "Passing positional arguments {remapped_names} to {func_name}()"
+        " is deprecated. Change them to keyword args."
+    )
+
+    def decorator(func):
+        if isinstance(func, type):
+            raise TypeError(
+                "@deprecate_posargs cannot be applied to a class."
+                " (Apply it to the __init__ method.)"
+            )
+        if isinstance(func, classmethod):
+            raise TypeError("Apply @classmethod before @deprecate_posargs.")
+        if isinstance(func, staticmethod):
+            raise TypeError("Apply @staticmethod before @deprecate_posargs.")
+
+        params = inspect.signature(func).parameters
+        num_by_kind = Counter(param.kind for param in params.values())
+
+        if num_by_kind[inspect.Parameter.VAR_POSITIONAL] > 0:
+            raise TypeError(
+                "@deprecate_posargs() cannot be used with variable positional `*args`."
+            )
+
+        num_positional_params = (
+            num_by_kind[inspect.Parameter.POSITIONAL_ONLY]
+            + num_by_kind[inspect.Parameter.POSITIONAL_OR_KEYWORD]
+        )
+        num_keyword_only_params = num_by_kind[inspect.Parameter.KEYWORD_ONLY]
+        if num_keyword_only_params < 1:
+            raise TypeError(
+                "@deprecate_posargs() requires at least one keyword-only parameter"
+                " (after a `*` entry in the parameters list)."
+            )
+        if any(
+            name not in params or params[name].kind != inspect.Parameter.KEYWORD_ONLY
+            for name in remappable_names
+        ):
+            raise TypeError(
+                "@deprecate_posargs() remappable_names must"
+                " all be keyword-only parameters."
+            )
+
+        num_remappable_args = len(remappable_names)
+        max_positional_args = num_positional_params + num_remappable_args
+
+        func_name = func.__name__
+        if func_name == "__init__":
+            # In the warning, show "ClassName(...)" rather than "__init__(...)".
+            # The class isn't defined yet, but its name is in __qualname__.
+            # Some examples: "ClassName.__init__", "Nested.ClassName.__init__",
+            # "MyTests.test_case.<locals>.ClassName.__init__".
+            local_name = func.__qualname__.rsplit("<locals>.", 1)[-1]
+            class_name = local_name.replace(".__init__", "")
+            func_name = class_name
+
+        def remap_deprecated_args(args, kwargs):
+            """
+            Move deprecated positional args to kwargs and issue a warning.
+            Returns updated (args, kwargs).
```

Comment:
PEP-257 nitpick, verbs should be written as a command in functions/methods docstrings:
```suggestion
            Return updated (args, kwargs).
```